Don't stop systemd-{networkd,resolved,udevd} on config switch

antifuchs opened the pull request:
These daemons should not be stopped, as they're foundational to a proper functioning of the system. When switching configurations, they only need a restart instead of that stop/start cycle.

This is a continuation of #359724 and the closed #372191 (which tried to do what this PR does, in a more general way).

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

antifuchs committed:
Don't stop systemd-{networkd,resolved,udevd} on config switch

These daemons should not be stopped, as they're foundational to a
proper functioning of the system. When switching configurations, they
only need a restart instead of that stop/start cycle.

ElvishJerricco reviewed (commented):
While we're at it, I wonder if the `restartTriggers` in `udevd` and `resolved` should be `reloadTriggers` instead. `man systemd-resolved` indicates that should work for `resolved`, but it's less clear for `udevd`. And it looks like the `networkd` module is probably already differentiating the two types of triggers correctly.

antifuchs committed:
Make systemd-resolved's config file a reload trigger

It is documented to re-read its configuration file upon reload, so
we can simply reload it instead of restarting the whole daemon.

antifuchs commented:
> While we're at it, I wonder if the `restartTriggers` in `udevd` and `resolved` should be `reloadTriggers` instead. `man systemd-resolved` indicates that should work for `resolved`, but it's less clear for `udevd`. And it looks like the `networkd` module is probably already differentiating the two types of triggers correctly.

Yeah, I'm not sure about udevd either - apparently it requires a custom invocation via `udevadm` to reload the triggers... but resolved seems very safe. I'll do that and leave udevd to a braver soul to investigate!

antifuchs commented:
Built a system configuration to test it, and here's the pertinent dry-activate lines:

```
would stop the following units: atuin.service, buildbot-master.service, [...], systemd-tmpfiles-resetup.service
would NOT stop the following changed units: libvirt-guests.service, libvirtd.service, virtlogd.service
would activate the configuration...
would reload the following units: buildbot-worker.service, dbus.service
would restart the following units: nix-daemon.service, polkit.service, systemd-networkd.service, systemd-resolved.service, systemd-udevd.service
would start the following units: atuin.service, buildbot-master.service, [...], systemd-tmpfiles-resetup.service
```

Notice the restart of networkd and resolved! Think we are on the winning path (:
